src/util/cilfacade.ml

```diff
@@ -80,71 +80,45 @@ let end_basic_blocks f =
   let thisVisitor = new allBBVisitor in
   visitCilFileSameGlobals thisVisitor f
 
+class removeLabelsVisitor = object
+  inherit nopCilVisitor
+
+  val nestingDepth = ref 0
+
+  method! vstmt s =
+    let new_labels = List.filter_map (function Label(str,loc,b) -> None | x -> Some x) s.labels in
+    match s.skind with
+    | Loop _ -> nestingDepth := !nestingDepth+1; ChangeDoChildrenPost({s with labels = new_labels}, fun x -> nestingDepth := !nestingDepth-1; x)
+    | Continue loc -> if !nestingDepth = 0 then
+        ChangeTo({s with labels = new_labels; skind = Break loc})
+      else
+        ChangeTo({s with labels = new_labels})
+    | _ -> ChangeDoChildrenPost({s with labels = new_labels}, Fun.id)
+end
+
 class loopUnrollingVisitor = object
+  (* Labels are simply handled by removing them. It is always correct to simply jump into the non-unrolled part of the loop!          *)
+  (* The price we need to pay for this simpler handling of labels and still wanting to benefit from unrolling nested loops is that    *)
+  (* we need to put our unrolled code into fake while(1) loops such that breaks have something to break out of, add artificial breaks *)
+  (* at the end of loops, and replacing (top level!) continues with breaks. *)
+  (* This is still a lot more intuitive than complicated handling of gotos as would be required when doing it after the unrolling     *)
   inherit nopCilVisitor
+    
   method! vstmt s =
-    (*
-    All the duplicated stmts need to be re-created or the sid will fail, so we create a copy of st, which is physically unequal to it.
-    A new sid is  computed later (as seen in http://goblint.in.tum.de/assets/goblint-cil/api/Cil.html#TYPEstmt), as long as we perform this copies.
-    *)
-    let newsid st = { st with sid = st.sid } in
-    let rec newsid_stmt st =
-      let mkb stml = mkBlock (List.map newsid_stmt stml) in
-      match st.skind with
-      | Instr _ -> newsid st
-      | If(e,b1,b2,l1,l2) -> mkStmt(If(e,(mkb b1.bstmts),(mkb b2.bstmts),l1,l2))
-      | Loop(bl,l1,l2,s1,s2) ->
-        let create_stmt = mkStmt(Loop((mkb bl.bstmts),l1,l2,s1,s2)) in
-        create_stmt.labels <- st.labels;
-        create_stmt
-      | Block(bl) -> mkStmt(Block(mkb bl.bstmts))
-      | Switch(e,bl,stl,l1,l2) -> mkStmt(Switch(e,(mkb bl.bstmts),(List.map newsid_stmt stl),l1,l2))
-      | _ -> newsid st in
-    let create_shallow_copies bl = List.map newsid_stmt bl in
+    let labelsRemover = new removeLabelsVisitor in
```

Comment:
This could instead be a `val` in the `loopUnrollingVisitor` class. Currently it allocates a new object for each visited unconditionally.